### ISSUE TITLE
Fix S1643: check if expression is concatenation

### ISSUE THREAD

**Issue by koczkatamas** [SonarSource/sonar-csharp#1138]
### Description

The following code triggers a S1643 `Strings should not be concatenated using '+' in a loop` error:

```csharp
string fileContent = null;
foreach (var item in items)
  if (item.requiresFile)
     fileContent = fileContent ?? FileSystem.ReadAllText(fn);
```

I think the root cause of this issue is that you never check that the binary expression (in this case `x ?? y`) uses the string concatenation operator `+` or not.

I am not 100% sure, but I think a check is missing after this:
https://github.com/SonarSource/sonar-csharp/blob/b8a7415a5cee73685924cf92b7ee3ed7c3bbdb16/sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Rules/StringConcatenationInLoopBase.cs#L64

### Repro steps

Try to analize this code:
```csharp
string fileContent = null;
foreach (var item in new[]{ false, true })
  if (item)
     fileContent = fileContent ?? FileSystem.ReadAllText(fn);
```

### Expected behavior

SonarQube does not warn about S1643.

### Actual behavior

SonarQube DOES warn about S1643.

### Known workarounds

No workaround is known at the moment.

### Related information

* SonarC# Version: `SonarQube Scanner for MSBuild 4.0.2`
* Visual Studio Version: `2017`
* If running through the Scanner for MSBuild, its version: `15.5.180.51428`

**Comment by Evangelink** [comment]
Hi @koczkatamas,

That's indeed a False Positive.
